### PR TITLE
Use same GPU stream for all kernels

### PR DESCRIPTION
**Pull request by MrBurmark**
# Use the same GPU stream for all kernels

Use a specific GPU stream for all Cuda/Hip kernels. This is done by using the same resource for all kernels. By default this is the RAJA default stream, but can be changed to stream 0 with the `--gpu_stream_0` command line argument.

- This PR is a feature
- It does the following:
  - Adds GPU stream control as requested in #294

### CODE REVIEW

**Comment by MrBurmark** [comment]
One thing I didn't think too deeply about while making this PR is what to do with calls to cudaMemcpy (non-async). I believe those API calls run on stream 0 which isn't a correctness problem as the RAJA default stream is synchronous with stream 0, but it isn't really in the spirit of the change in this PR either. I'm thinking about adding a resource argument to some of our memory helper functions to work around this.

**Comment by MrBurmark** [comment]
Most of them don't matter, but there are some that are used in the timing loop in reduction kernels. I've been trying it out and it looks like there is a performance penalty for mixing streams. I'm going to rewrite those memory calls to explicitly call cuda|hipMemcpyAsync and streamSynchronize.

**Comment by MrBurmark** [comment]
I made the change for the kernels using memcpy and synchronize calls in the timed portion. They now use memcpyAsync+streamSynchronize or streamSynchronize to use their stream.

File: src/algorithm/MEMCPY-Cuda.cpp [hunk 1]
```diff
@@ -105,7 +107,7 @@ void MEMCPY::runCudaVariantBlock(VariantID vid)
     for (RepIndex_type irep = 0; irep < run_reps; ++irep) {
 
       const size_t grid_size = RAJA_DIVIDE_CEILING_INT(iend, block_size);
-      memcpy<block_size><<<grid_size, block_size>>>(
+      memcpy<block_size><<<grid_size, block_size, 0, res.get_stream()>>>(
```
Comment (rhornung67) [inline src/algorithm/MEMCPY-Cuda.cpp:110]: What is the '0' arg for?

Comment (MrBurmark) [reply src/algorithm/MEMCPY-Cuda.cpp:110]: shared memory

**Review by rhornung67 — APPROVED** [review]
A lot of changes for something we should have thought about earlier, huh? 😄 Thank you for working through this.

File: src/algorithm/MEMCPY-Cuda.cpp (hunk 1 above)
Comment (artv3) [inline src/algorithm/MEMCPY-Cuda.cpp:110]:
```suggestion
      constexpr size_t shmem_sz = 0;
      memcpy<block_size><<<grid_size, block_size, 0, res.get_stream()>>>(
```

for clarity?

Comment (MrBurmark) [reply src/algorithm/MEMCPY-Cuda.cpp:110]: That's probably best practice, I can put that in if people would like that to be made more clear.

Comment (rhornung67) [reply src/algorithm/MEMCPY-Cuda.cpp:110]: I agree.

Comment (MrBurmark) [reply src/algorithm/MEMCPY-Cuda.cpp]: done

**Review by CRobeck — APPROVED** [review]

**Review by artv3 — APPROVED** [review]
Wonderful!

**Review by rhornung67 — APPROVED** [review]

**Comment by rhornung67** [comment]
@MrBurmark let's merge this one next when it gets through tioga CI

**Pull request merged**